ratter/cli/parser.py

```diff
@@ -0,0 +1,381 @@
+import argparse
+
+from abc import ABC, abstractstaticmethod
+from argparse import ArgumentError, ArgumentParser, Namespace
+from os.path import isfile, splitext
+
+from ratter import error
+from ratter.cli.util import multi_paragraph_wrap
+
+
+def parse_arguments() -> Namespace:
+    """Return the parsed and validated CLI arguments.
+
+    The parser is constructed and validated using the registered
+    `ArgumentGroupParser`s.
+
+    Each `ArgumentGroupParser::register` adds an
+    `argparse.ArgumentParser::add_argument_group` to the parser and returns the
+    parser.
+
+    Each `ArgumentGroupParser::validate` validates the current state of the
+    arguments, modifies them if need be, and returns the resulting arguments.
+    If there is invalidity within the arguments, then the method should raise
+    and `argparse.ArgumentError` with an appropriate message.
+
+    On error the program usage and the error message will be shown before
+    exiting.
+
+    """
+    parser: ArgumentParser = argparse.ArgumentParser(
+        prog="ratter",
+        description=multi_paragraph_wrap("""\
+            Parse a given Python 3 file to find the attributes used in each
+            function.
+
+            Given the regular expression <filter-string>, the results will only
+            include functions, classes, or methods whose name fully matches
+            regular expression.
+
+            Uses Python's regular expression syntax.
+        """),
+        epilog=multi_paragraph_wrap("""\
+            This is the epilogue, put a Suade thing here.
```

Comment:
What do you mean by this line?

Reply:
The gitlab fix (and other fixes in the mean time) have been added :)